File: src/main/java/queue/QueueDummy.java
package queue;

/**
 * Class representing a single node in a singly linked list.
 */
class ListNode {
    int val;        // Value stored in the node
    ListNode next;  // Pointer to the next node

    /**
     * Constructor to initialize a node with a value.
     *
     * @param val The value to store in the node.
     */
    public ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

/**
 * Class representing a queue implemented using a singly linked list.
 */
class QueueDummy {
    private ListNode left;  // Points to the front of the queue
    private ListNode right; // Points to the rear of the queue

    /**
     * Constructor to initialize an empty queue.
     */
    public QueueDummy() {
        this.left = null;  // Initialize the front pointer
        this.right = null; // Initialize the rear pointer
    }

    /**
     * Add an element to the end of the queue.
     *
     * @param val The value to be added to the queue.
     */
    public void enqueue(int val) {
        ListNode newNode = new ListNode(val); // Create a new node

        if (right != null) { // If the queue is not empty
            right.next = newNode; // Link the new node to the current rear
            right = newNode;      // Update the rear pointer to the new node
        } else { // If the queue is empty
            left = right = newNode; // Both pointers point to the new node
        }
    }

    /**
     * Remove and return the front element of the queue.
     *
     * @return The value of the removed element, or -1 if the queue is empty.
     */
    public int dequeue() {
        if (left == null) { // If the queue is empty
            System.out.println("Queue is empty!");
            return -1; // Indicate an empty queue
        }

        int val = left.val;     // Store the value of the front node
        left = left.next;       // Move the front pointer to the next node

        if (left == null) {     // If the queue becomes empty after removal
            right = null;       // Reset the rear pointer
        }
        return val;
    }

    /**
     * Print all elements in the queue from front to rear.
     */
    public void print() {
        ListNode cur = left; // Start from the front of the queue
        while (cur != null) {
            System.out.print(cur.val + " -> ");
            cur = cur.next;
        }
        System.out.println("None"); // Indicate the end of the queue
    }

   
}

File: python/linked_list/QueueDummy.py
class ListNode:
    """
    Class representing a single node in a singly linked list.
    """
    def __init__(self, val):
        self.val = val       # Value stored in the node
        self.next = None     # Pointer to the next node


class QueueDummy:
    """
    Class representing a queue implemented using a singly linked list
    with dummy nodes for easier management.
    """
    def __init__(self):
        """
        Initialize the queue with `left` and `right` pointers.
        Both pointers are initially set to None, representing an empty queue.
        """
        self.left = None  # Points to the front of the queue
        self.right = None # Points to the rear of the queue

    def enqueue(self, val):
        """
        Add an element to the end of the queue.
        :param val: The value to be added to the queue.
        """
        newNode = ListNode(val)  # Create a new node with the given value

        if self.right:  # If the queue is not empty
            self.right.next = newNode  # Link the new node to the current rear
            self.right = newNode       # Update the rear pointer to the new node
        else:  # If the queue is empty
            self.left = self.right = newNode  # Both pointers point to the new node

    def dequeue(self):
        """
        Remove and return the front element of the queue.
        :return: The value of the removed element, or None if the queue is empty.
        """
        if not self.left:  # If the queue is empty
            return None

        val = self.left.val  # Store the value of the front node
        self.left = self.left.next  # Move the front pointer to the next node

        if not self.left:  # If the queue is now empty after removal
            self.right = None  # Reset the rear pointer
        return val

    def print(self):
        """
        Print all elements in the queue from front to rear.
        """
        cur = self.left  # Start from the front of the queue
        while cur:
            print(cur.val, ' -> ', end="")  # Print each value followed by an arrow
            cur = cur.next
        print()  # New line after printing all elements

